Blake2b F precompile

pdyraga opened the pull request:
Here we present an implementation of Blake2b F compression function precompile based on the https://github.com/ethereum/EIPs/pull/2129

### Description

The precompile lives at the address `0x09` and should be called with a standard `staticcall` instruction. The expected input data structure is as follows:
```
[ 4 bytes for rounds ][ 64 bytes for h ][ 128 bytes for m ] \
[ 8 bytes for t_0 ][ 8 bytes for t_1 ][ 1 byte for f ]
```
where:
- `rounds` - number of rounds
- `h` - state vector
- `m` - message block vector
- `t_0, t_1` - offset counter
- `f` - final block indicator flag

There is a sample Truffle project prepared which can be used to play with this precompile:
https://github.com/pdyraga/f-precompile-call :
```
truffle migrate --reset --network local && truffle exec ./scripts/query-f.js --network local
```

### Istanbul hardfork

Since we hope to have this implementation included in the Istanbul hardfork, all control structures related to precompiles and hardfork have been added in this PR. Before you'll be able to use this precompile you need to enable Istanbul HF. One possibility is to use `genesis.json` and set `"istanbulBlock": 0` in `"config"` section.

### Sample call
```
 function F(uint32 rounds, bytes32[2] memory h, bytes32[4] memory m, uint64[2] memory t, bool f) public view returns (bytes32[2] memory) {
      bytes32[2] memory output;

      bytes memory args = abi.encodePacked(rounds, h[0], h[1], m[0], m[1], m[2], m[3], t[0], t[1], f);

      assembly {
            if iszero(staticcall(not(0), 0x09, add(args, 32), 0xd5, output, 0x40)) {
                revert(0, 0)
            }
      }

      return output;
    }

    function callF() public view returns (bytes32[2] memory) {
      uint32 rounds = 12;

      bytes32[2] memory h;
      h[0] = hex"b02406865b80cde9fec1e321559891be456158a6d71cf0eac1eb845c70d4a3d8"; 
      h[1] = hex"948e82cc975dcbd6984780bab7e236382fb585df5bc4b043cb3e866544413d92";
  
      bytes32[4] memory m;
      m[0] = hex"87868584838281808f8e8d8c8b8a898897969594939291909f9e9d9c9b9a9998";
      m[1] = hex"a7a6a5a4a3a2a1a0afaeadacabaaa9a8b7b6b5b4b3b2b1b0bfbebdbcbbbab9b8";
      m[2] = hex"c7c6c5c4c3c2c1c0cfcecdcccbcac9c8d7d6d5d4d3d2d1d0dfdedddcdbdad9d8";
      m[3] = hex"e7e6e5e4e3e2e1e0efeeedecebeae9e8f7f6f5f4f3f2f1f000fefdfcfbfaf9f8";
 
      uint64[2] memory t;
      t[0] = 383; // 000000000000017f
      t[1] = 0;	// 0000000000000000

      bool f = true; // 01

      return F(rounds, h, m, t, f);
    }
```

pdyraga committed:
Introduced map of precompiled Istanbul contracts

So far the precompiled Istanbul contracts list contains exactly the same
contracts as the one for Byzantium.

We'll work on adding a support for Blake2 F precompile in Istanbul in the
next commits.

Next to PrecompiledContractsIstanbul list, I introduced all the required
control structures for the fork.

pdyraga committed:
Added dependency to github.com/keep-network/blake2-f

pdyraga committed:
keep-network/blake2 dependency updated to the newest version

pdyraga committed:
Added Blake2b F function precompile contract

The contract calls keep-network/blake2 implementation of the F
compression function and lives under address 0x09. The called F function
is optimized for 64-bit WORDs.

Test cases were generated from test vectors from keep-network/blake2
library.

pdyraga committed:
Reverted mistakenly modified rootPath in vendor.json

pdyraga commented on core/vm/contracts.go:
```diff
@@ -358,3 +375,49 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+type blake2F struct{}
+
+func (c *blake2F) RequiredGas(input []byte) uint64 {
+	return 1 //TODO: implement
```
This still needs to be implemented but we need to decide first what's the strategy.

pdyraga replied:
Benchmarked on my Mac:
```
BenchmarkPrecompiledEcrecover/-Gas=3000-8                  10000            180309 ns/op
```

```
BenchmarkPrecompiledSha256/128-Gas=108-8                 2000000               614 ns/op
```

```
BenchmarkPrecompiledBlake2F/testVectors2bX_0-Gas=1-8             3000000               382 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_3-Gas=1-8             3000000               383 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_70-Gas=1-8            5000000               382 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_140-Gas=1-8           3000000               374 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_230-Gas=1-8           5000000               379 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_300-Gas=1-8           5000000               387 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_370-Gas=1-8           5000000               462 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_440-Gas=1-8           5000000               477 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_510-Gas=1-8           3000000               421 ns/op
BenchmarkPrecompiledBlake2F/testVectors2bX_580-Gas=1-8           5000000               376 ns/op
```

pdyraga replied:
Continuing the thread in the [EIP PR](https://github.com/keep-network/EIPs/pull/2/files#r295576690)

hadv commented on core/vm/contracts.go:
```diff
@@ -18,9 +18,12 @@ package vm
 	"math/big"
 
+	"github.com/keep-network/blake2"
+
```
nit: no need blank line here ;)

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +375,49 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+	h[4] = binary.BigEndian.Uint64(input[32:40])
+	h[5] = binary.BigEndian.Uint64(input[40:48])
+	h[6] = binary.BigEndian.Uint64(input[48:56])
+	h[7] = binary.BigEndian.Uint64(input[56:64])
```
The EIP needs to specify how to handle cases where the input data is too short. Other precompiles typially return an error in that case. It's very important that such decisions are not left "as implementation details", since it's consensus-critical behaviour. 

In this case, the client woud `panic` if  given an input shorter than `64` bytes, afaict

mhluongo replied:
I'll add it to the EIP spec, waiting for the draft and competing PRs to be resolved

pdyraga replied:
We now expect the input to be exactly 213 bytes long. 

The PR was opened as an initial implementation before we figure out some details. I should probably open it in a Draft mode.

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +375,49 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+		f = true
+	}
+
+	rounds := binary.BigEndian.Uint32(input[316:320])
```
actually, `panic` on shorter than  `320` bytes in length

pdyraga replied:
Yep, it was wrong. We now expect the input to be exactly 213 bytes long.

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +375,49 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+type blake2F struct{}
+
+func (c *blake2F) RequiredGas(input []byte) uint64 {
+	return 1 //TODO: implement
```
Even if you don't quite know what constants to use, perhaps you have an idea about the general pricing scheme? Like, is it `rounds * K *input_size` ? Would be good to have _something_, even if we determine the constants later. It would make it easier for others people to help tune the constants.

mhluongo replied:
The benchmarks so far lead me to believe that a round is ~ 1 gas (!), and that this precompile will be dominated by calling cost. We can provide a more robust set of benchmarks that including input size as a floating param.

More discussion [here](https://github.com/keep-network/EIPs/pull/2/files#discussion_r295576690)

mhluongo replied:
I believe the methodology of pricing relative to `ecrecover` is sound. Once we have a wider range of inputs on benchmarks this will be straightforward.

pdyraga replied:
I opened this PR in an early-mode before we had a chance to discuss benchmarks and gas price. Based on what we know so far, and using the same methodology as for the [EIP-1108](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1108.md) it seems that it should be [1 gas/round](https://github.com/ethereum/EIPs/pull/2129#discussion_r296223207).

I was hoping to remove this TODO and put a final value here after we agree on the gas [in the EIP](https://github.com/ethereum/EIPs/pull/2129). I am not sure how the process should look here.

pdyraga replied:
I set the gas price to 1/round based on our benchmarks.

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +375,49 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+	t[1] = binary.BigEndian.Uint64(input[248:256])
+
+	var f bool
+	if input[287] == 1 {
```
Interesting -- this right here checks if the `f` is _strictly_ `b 0000 0001`. So if it's `b 1111 1111`, it would be set to `false`. I think it would be more sensible to look purely at the lsb (and, naturally, to have that specified in the EIP).

pdyraga replied:
I updated the check to:
```
if input[208] == 0x00000001 {
	f = true
}
```
I am not sure if it is what you meant by looking at the least significant bit.

holiman replied:
I meant something like
```
if (input[208] & 0x1) != 0 {
	f = true
}
```
There are basically three variants:
1. Check if `input[208] != 0`. So `b 1111 1110` counts as `true`
2. Check if `input[208] == 1` (what you do now). So `b 1111 1111` counts as `false`
3. Check if only least significant bit. 

My order of preference would be 3, 1, 2. Another option would be to do _exactly_ what solidity does, whatever that is, to resolve an input parameter into a `bool`. @axic what would solidity do ?

holiman replied:
I just experimented a bit with solidity. Afaict, for `boolean`, the rule seems to be `nonzero = true` (variant 1). So if we want to follow the same logic, the check would be 
```
f := (input[208] != 0)
```

pdyraga replied:
Good point, updated as requested.

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +375,49 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+
+	rounds := binary.BigEndian.Uint32(input[316:320])
+
+	blake2.F(&h, m, t, f, int(rounds))
```
Is it really safe to cast a `uint32` to `int` -- what about 32-bit systems? Will it become negative on `0xffff ffff`?

pdyraga replied:
I think it is safe but... better be safe than sorry. I will update `blake2.F` to accept the number of rounds as `int32`.

pdyraga replied:
We no longer cast to `int` - the number of rounds is `uint32` everywhere now.

holiman commented:
Why is all the padding needed? The caller sends along `320` bytes, but the method totally ignores bytes `192 - 216` , `256 - 286`, `288 - 316` ?

pdyraga commented on core/vm/contracts.go:
```diff
@@ -358,3 +375,49 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+	return 1 //TODO: implement
+}
+
+func (c *blake2F) Run(input []byte) ([]byte, error) {
```
> Why is all the padding needed? The caller sends along 320 bytes, but the method totally ignores bytes 192 - 216 , 256 - 286, 288 - 316 ?

@holiman I thought we always have to use 256-bit boxes for individual arguments to `staticcall`. 🤔 

I was experimenting with `abi.encodePacked` but I've run into some problems with `staticcall` and memory handling. Here is what I do:
```
    function F(bytes32[2] memory h, bytes32[4] memory m, uint64[2] memory t, bool f, uint32 rounds) public view returns (bytes32[2] memory) {
      bytes32[2] memory output;

      bytes memory args = abi.encodePacked(h[0], h[1], m[0], m[1], m[2], m[3], t[0], t[1], f, rounds);

      assembly {
            if iszero(staticcall(not(0), 0x09, args, 0xd5, output, 0x40)) {
                revert(0, 0)
            }
      }
      
      return output;
    }

    function callF() public view returns (bytes32[2] memory) {
      bytes32[2] memory h;
      h[0] = hex"6a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f2"; 
      h[1] = hex"510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179";
  
      bytes32[4] memory m;
      m[0] = hex"278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad";
      m[1] = hex"0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da";
      m[2] = hex"0000000000000000000000000000000000000000000000000000000000000000";
      m[3] = hex"0000000000000000000000000000000000000000000000000000000000000000";
 
      uint64[2] memory t;
      t[0] = 18446744073709551552;  // ffffffffffffffc0
      t[1] = 18446744073709551615;	// ffffffffffffffff

      bool f = true;

      uint32 rounds = 12;

      return F(h, m, t, f, rounds);
    }
```

And here is what is received in the precompile:
```
func (c *blake2F) Run(input []byte) ([]byte, error) {
	log.Info("F precompile", "len", len(input), "input", input, "input_hex", hex.EncodeToString(input))
```


> INFO [06-24|18:47:05.016] F precompile                             len=213 input="[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 213 106 9 230 39 243 188 201 9 187 103 174 132 132 202 167 59 60 110 243 114 254 148 184 43 165 79 245 58 95 29 54 242 81 14 82 127 173 230 130 209 155 5 104 140 43 62 108 31 31 131 217 171 251 65 189 107 91 224 205 25 19 126 33 121 39 132 0 52 14 107 5 197 117 37 146 197 44 79 18 18 146 234 252 81 204 1 169 151 180 174 209 52 9 41 143 173 13 153 204 200 247 100 83 217 179 102 30 92 77 50 93 119 81 20 125 177 116 137 4 109 22 130 213 14 239 164 161 218 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]" input_hex=00000000000000000000000000000000000000000000000000000000000000d56a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f2510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

My `args` are left-padded with `00000000000000000000000000000000000000000000000000000000000000d5` and for this reason they are truncated on the right side.

I checked and `abi.encodePacked` returns a correct value:

> 0x6a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f2510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 0000000000000000000000ffffffffffffffc0ffffffffffffffff010000000c
            
Something wrong happens later and I have no idea yet what it is.

holiman replied:
I think it's that `bytes` is a dynamic array, 

> Arrays have a length member that contains their number of elements. The length of memory arrays is fixed (but dynamic, i.e. it can depend on runtime parameters) once they are created.

So the `0x00...d5` is probably the length of the `bytes` array. @axic ?

pdyraga replied:
It is highly possible it is the lenght of the array but if so, it has to be added somewhere inside `staticcall`, maybe when we push args to client's memory because I do not see this prefix when I emit an event with the result of `abi.encodePacked`.

axic replied:
You are using assembly above which gives a low level access to everything. One must be aware how types in memory are encoded. In `staticcall(..., args, ...)` the parameter `args` points to the memory area where that data resides, but in the case of `bytes memory` it is prefixed with the length.

In short, you need to use `staticcall(..., add(args, 32), ...)`.

pdyraga replied:
You were right, `add` helped. Thank you @axic and @holiman.

pdyraga committed:
Accept abi.encodePacked parameters to F precompile

We decided to switch the behaviour of F precompile to accept
abi.encodePacked parameters instead of left-padded ones (for values
smaller than 32 bytes).

pdyraga committed:
Allow only 213-bytes input in F precompile

pdyraga committed:
Set Blake2b F precompile price to 1gas/round

pdyraga committed:
Reordered imports

No changes in what modules are imported. I've just removed the empty
line and auto-formatted imports.

pdyraga commented:
> Why is all the padding needed? The caller sends along 320 bytes, but the method totally ignores bytes 192 - 216 , 256 - 286, 288 - 316 ?

We no longer have padding. We now accept exactly 213 bytes of abi-encoded parameters.

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +374,68 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+type blake2F struct{}
+
+func (c *blake2F) RequiredGas(input []byte) uint64 {
+	if len(input) != 213 {
```
If input larger than `213` is used, this will throw. If that's the intended behaviour, then the EIP must also be updated to explicitly state that only exactly `213` bytes is accepted.

pdyraga replied:
Yes, my intention was to update the EIP to state it explicitly that exactly `213` bytes is accepted. I am not sure if there is any use case for passing more bytes and ignoring some of them. I personally think it's less error-prone to expect the exact length and fail when the length does not match.

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +374,68 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+type blake2F struct{}
+
+func (c *blake2F) RequiredGas(input []byte) uint64 {
+	if len(input) != 213 {
```
```suggestion
	if len(input) != blake2FInputLength {
```

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +374,68 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+}
+
+var (
+	blake2FInputLength = 213
```
this should be a `const`, no?

pdyraga replied:
Yeah, it could be:
```
const blake2FInputLength = 213

var errBlake2FIncorrectInputLength = errors.New(
	"input length for Blake2 F precompile should be exactly 213 bytes",
)
```

but thought it's better to initialize them in one block, similarly how it's done above the `bn256Pairing` struct definition. I wasn't sure about the conventions used in the project.

Please let me know what's the preference here and I'll alter it.

holiman replied:
I think `const` is preferrable, rather than calling it a `var` just for the sake of indentation...

pdyraga replied:
Good point, updated.

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +374,68 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+)
+
+func (c *blake2F) Run(input []byte) ([]byte, error) {
+	if len(input) != 213 {
```
```suggestion
	if len(input) != blake2FInputLength {
```

holiman commented on core/vm/contracts.go:
```diff
@@ -358,3 +374,68 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+
+func (c *blake2F) Run(input []byte) ([]byte, error) {
+	if len(input) != 213 {
+		return nil, errBlake2FIncorrectInputLength
```
If it's only used in this place, might as well define the error inline
```suggestion
		return nil, errors.New("invalid input length for Blake2 F precompile")
```

pdyraga replied:
It's also used in tests, see `TestPrecompileBlake2FMalformedInput`.

pdyraga committed:
Use blake2FInputLength constant to check the input length

Co-Authored-By: Martin Holst Swende <martin@swende.se>

iikirilov commented on core/vm/contracts.go:
```diff
@@ -358,3 +374,68 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+	t[1] = binary.BigEndian.Uint64(input[200:208])
+
+	var f bool
+	if input[208] == 0x00000001 {
```
```suggestion
	if input[208] != 0x00000000 {
```

pdyraga replied:
This will be checked in a different way, just `!= 0 `

pdyraga committed:
Made blake2FInputLength a constant

pdyraga committed:
Updaded keep-network/blake2b reference to the newest version

We now point to keep-network/blake2b/compression to include only what is
really needed - F compression function

pdyraga committed:
Switched to the new interface of compression.F, updated test vectors

compression.F accepts `m` parameter as an array of uint64 instead of
byte slice now. I have also improved test vectors to cover various cases
of rounds parameter.

pdyraga committed:
Accept rounds parameter as the first one to F precompile

Gas price depends solely on the number of rounds. Having this parameter
as the first one lets to evaluate gas price easier.

pdyraga committed:
nonzero f bytes are considered as 'true'

pdyraga committed:
Simplified condition assigning a value to `f` parameter

pdyraga committed:
Gas calculation function reads number of rounds from the first 4 bytes

I made a mistake when refactoring `Run(input []byte)` to do the same and
not updated `RequiredGas(input []byte)`. Fixing my mistake now.

pdyraga commented:
All comments addressed. I still need to do some more extensive tests but I think it's ready for another round of review.

pdyraga committed:
Updated blake2b compression lib reference to point to v1.0 release

Instead of pointing to a commit with no official tag, we now point to an
exact release number.

pdyraga commented:
Tested with https://github.com/pdyraga/f-precompile-call, works as expected:
```
$ truffle migrate --reset --network local && truffle exec ./scripts/query-f.js --network local
Compiling ./contracts/PrecompileCall.sol...
Writing artifacts to ./build/contracts

(...)

Using network 'local'.

querying...
0xccfc282ed60927145b46f8d0fa97afd07010c51d20821e9748923ea42a37a0fa,0x0609a13be7c1e14b6e10a4b63d85d1d56d3d370d80f97b0a61a4f22ed6462dee
```

daira commented on core/vm/contracts.go:
```diff
@@ -358,3 +374,58 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
+	for i := 0; i < 8; i++ {
+		offset := i * 8
+		binary.BigEndian.PutUint64(output[offset:offset+8], h[i])
+	}
```
This code may be using the wrong byte order, as discussed starting at https://github.com/ethereum/EIPs/pull/2129#discussion_r300456654 .

holiman commented:
Is this PR now 'final' ? If so, could you open a PR against go-ethereum proper?

pdyraga commented:
@holiman I need to update EIP with the requested test vectors and double-check possible byte ordering problems. Please expect a PR against `go-ethereum` this week.

holiman commented:
How is it going? Istanbul is not far off now, and we're trying to get things in as soon as possible, to allow more time for testing and test creation.

pdyraga commented:
> How is it going? Istanbul is not far off now, and we're trying to get things in as soon as possible, to allow more time for testing and test creation.

Working on it today, sorry for the delay.

karalabe commented on vendor/vendor.json:
```diff
@@ -273,6 +273,14 @@
+			"revision": "02884c605a94951d05165fd56c69d6239f63b3d1",
+			"revisionTime": "2019-07-02T09:51:07Z",
+			"version": "v1.0",
+			"versionExact": "v1.0"
```
Since this is actually code taken verbatim from Go (as far as I understand it), please move it to `crypto/blake2b` inside the repo. I don't like the idea that we have a middleman repo in between for code that's copy pasted from elsewhere anyway.

pdyraga replied:
This is not entirely copy-paste. Go code does not expose `F` compression function, we had to modify it to allow for an arbitrary number of rounds, have the interface compatible with what we expect from the precompile and removed some pieces that are not really a part of compression but rather hashing. Please compare:
https://github.com/keep-network/blake2b/blob/master/compression/f.go#L39
https://github.com/golang/crypto/blob/master/blake2b/blake2b_generic.go#L30

karalabe commented on vendor/vendor.json:
```diff
@@ -965,5 +973,5 @@
 		}
 	],
-	"rootPath": "github.com/ethereum/go-ethereum"
+	"rootPath": "github.com/keep-network/go-ethereum"
```
Take care to fix this if you open a PR upstream.

pdyraga committed:
Accept h,m,t bytes for BLAKE2b precompile as little-endian

pdyraga committed:
Replaced generated test vectors with one from BLAKE2b RFC

BLAKE2b RFC Appendix A specifies test vector that we have implemented
here.

See https://tools.ietf.org/html/rfc7693#appendix-A

For now, we leave blake2FTests are slice because we'll probably add
more test vectors there.

pdyraga committed:
Strict validation of the final block indicator flag

The final block indicator flag for F precompile must be either 0 or 1.

holiman commented on core/vm/contracts.go:
```diff
@@ -389,15 +389,24 @@ func (c *blake2F) RequiredGas(input []byte) uint64 {
 	if len(input) != blake2FInputLength {
 		return nil, errBlake2FIncorrectInputLength
 	}
+	if input[212] != blake2FNonFinalBlockBytes && input[212] != blake2FFinalBlockBytes {
```
Now you changed it to be super strict. Is that in line with the EIP? It wasn't when I checked yesterday, IIRC

holiman replied:
Yes, the EIP still says

> The boolean f parameter is considered as true if at least 1 bit out of the 8 bits is set.

pdyraga replied:
I am going to update EIP in an ~hour. I should probably push the EIP change first, sorry. Changed based on [this](https://github.com/ethereum/EIPs/issues/152#issuecomment-508494194) discussion.

pdyraga committed:
Decode test vector with Hex2Bytes before modifying it

holiman commented:
@pdyraga please file this PR against go-ethereum, and let's do the remaining the work there. 

We're getting to a point where we are cutting very close to deadlines. 

So just open a PR, that's enough if you don't have more time right now. If you _do_ have some more time, it'd be great if you also rebase it against current `master` (and if you have even more time, if you also squash the commits).

Otherwise, one if us will have to take this branch, rebase it, and file it ourselves (if it eventually get's squashed, it will be better if you are the PR submitter, so you get commit-credi).

pdyraga commented:
I'll do it today @holiman 

I am currently working on the last test vectors for the EIP. Should have both ready today.

holiman commented:
Great!
But just so you know - having the final test vectors is good, but if you don't have time for both test vectors and upstream PR, then I'd prioritize the PR.

pdyraga commented:
@holiman Regarding squashing commits - is the preferred approach to have just one commit in the PR or did you mean that I should just clean up the history a little bit?

pdyraga commented:
Thanks for all the comments. Closing here, and opening on `go-ethereum` soon.
